File: Amazon Price Monitor (iOS)/AmazonPriceMonitor/Models/CartItem.swift
//
//  CartItem.swift
//  Amazon Price Monitor
//

import Foundation

struct PriceEntry: Codable, Identifiable {
    var id: String { date }
    let date: String
    let price: Double
}

struct CartItem: Codable, Identifiable {
    let id: String // ASIN
    var title: String
    var imageURL: String?
    var history: [PriceEntry]

    var currentPrice: Double? {
        history.last?.price
    }

    var lowestPrice: Double? {
        history.map(\.price).min()
    }

    var highestPrice: Double? {
        history.map(\.price).max()
    }

    var priceChangePercent: Double? {
        guard history.count >= 2,
              let first = history.first?.price,
              let last = history.last?.price,
              first > 0 else { return nil }
        return ((last - first) / first) * 100
    }

    enum Trend {
        case up, down, stable
    }

    var trend: Trend {
        guard let change = priceChangePercent else { return .stable }
        if change < -0.01 { return .down }
        if change > 0.01 { return .up }
        return .stable
    }
}

extension Double {
    /// Price formatted with the user's locale grouping/decimals, without the currency
    /// symbol (the "R$" prefix is rendered separately in the UI). e.g. "8.799,00" / "8,799.00".
    var priceValue: String {
        let formatter = NumberFormatter()
        formatter.numberStyle = .decimal
        formatter.minimumFractionDigits = 2
        formatter.maximumFractionDigits = 2
        return formatter.string(from: NSNumber(value: self)) ?? String(format: "%.2f", self)
    }
}

File: Amazon Price Monitor (Android)/app/src/main/java/com/vacantaurora/pricemonitor/model/CartItem.kt
package com.vacantaurora.pricemonitor.model

import kotlinx.serialization.Serializable
import java.text.NumberFormat

@Serializable
data class PriceEntry(
    val date: String, // ISO-8601 instant string
    val price: Double,
) {
    val id: String get() = date
}

@Serializable
data class CartItem(
    val id: String, // ASIN
    val title: String,
    val imageURL: String? = null,
    val history: List<PriceEntry> = emptyList(),
) {
    val currentPrice: Double? get() = history.lastOrNull()?.price

    val lowestPrice: Double? get() = history.minOfOrNull { it.price }

    val highestPrice: Double? get() = history.maxOfOrNull { it.price }

    val priceChangePercent: Double?
        get() {
            if (history.size < 2) return null
            val first = history.first().price
            val last = history.last().price
            if (first <= 0) return null
            return ((last - first) / first) * 100
        }

    enum class Trend { UP, DOWN, STABLE }

    val trend: Trend
        get() {
            val change = priceChangePercent ?: return Trend.STABLE
            return when {
                change < -0.01 -> Trend.DOWN
                change > 0.01 -> Trend.UP
                else -> Trend.STABLE
            }
        }
}

/**
 * Price formatted with the user's locale grouping/decimals, without the currency
 * symbol (the "R$" prefix is rendered separately in the UI). e.g. "8.799,00" / "8,799.00".
 */
fun Double.priceValue(): String {
    val formatter = NumberFormat.getNumberInstance().apply {
        minimumFractionDigits = 2
        maximumFractionDigits = 2
    }
    return formatter.format(this)
}
